From a8135d2dc4fe170bdb2d2fb23f3fbd054af55429 Mon Sep 17 00:00:00 2001
From: Simon Busch <morphis@gravedo.de>
Date: Fri, 22 Aug 2014 17:00:13 +0200
Subject: [PATCH] Bring back old setRenderWithoutShowing API method

API method was removed once ago because of not applying to the new rendering architecture
but we need it still for a specific use case where things are still working with it.

Upstream-Status: Inappropiate (bringing back old API functionality)
---
 src/quick/items/qquickwindow.cpp | 18 +++++++++++++++++-
 src/quick/items/qquickwindow_p.h |  3 +++
 2 files changed, 20 insertions(+), 1 deletion(-)

diff --git a/src/quick/items/qquickwindow.cpp b/src/quick/items/qquickwindow.cpp
index 5320219..de89e0a 100644
--- a/src/quick/items/qquickwindow.cpp
+++ b/src/quick/items/qquickwindow.cpp
@@ -268,6 +268,20 @@ void QQuickWindowPrivate::polishItems()
     updateFocusItemTransform();
 }
 
+void QQuickWindowPrivate::setRenderWithoutShowing(bool enable)
+{
+    if (enable == renderWithoutShowing)
+        return;
+
+    Q_Q(QQuickWindow);
+    renderWithoutShowing = enable;
+
+    if (enable)
+        windowManager->show(q);
+    else
+        windowManager->hide(q);
+}
+
 /*!
  * Schedules the window to render another frame.
  *
@@ -380,6 +394,7 @@ QQuickWindowPrivate::QQuickWindowPrivate()
 #endif
     , touchMouseId(-1)
     , touchMousePressTimestamp(0)
+    , renderWithoutShowing(false)
     , dirtyItemList(0)
     , context(0)
     , renderer(0)
@@ -2214,7 +2229,8 @@ void QQuickWindowPrivate::data_clear(QQmlListProperty<QObject> *property)
 bool QQuickWindowPrivate::isRenderable() const
 {
     Q_Q(const QQuickWindow);
-    return q->isExposed() && q->isVisible() && q->geometry().isValid();
+    return (q->isExposed() && q->isVisible() && q->geometry().isValid()) ||
+        (renderWithoutShowing && platformWindow);
 }
 
 /*!
diff --git a/src/quick/items/qquickwindow_p.h b/src/quick/items/qquickwindow_p.h
index 418633b..e0b6f5a 100644
--- a/src/quick/items/qquickwindow_p.h
+++ b/src/quick/items/qquickwindow_p.h
@@ -180,6 +180,9 @@ public:
 
     bool isRenderable() const;
 
+    bool renderWithoutShowing;
+    void setRenderWithoutShowing(bool enabled);
+
     QQuickItem::UpdatePaintNodeData updatePaintNodeData;
 
     QQuickItem *dirtyItemList;
-- 
1.9.1

